### ISSUE TITLE
Move to Typescript

### ISSUE THREAD

**Issue by lil5** [balzack/databag#89]
I'd like to use this project in a different project, I'd need to build an sdk based off the app/mobile/src/api files. Even then I'd be coding in the dark without types. A good first step would be to move to Typescript.

**Comment by balzack** [comment]
yes, I have some regret not starting off with typescript and have been postponing the refactor effort. I've added it to my backlog list for tracking, but I have a few features I've committed to first.

**Comment by lil5** [comment]
If it's alright with you I'd like to move to strict = false first, this makes most JavaScript okay for TypeScript. 
It allows for untyped code and missing types. 

That way the types can follow organically. 

I'd like to help out with this change. 

Are your unmerged branches very different from main?

**Comment by balzack** [comment]
I currently only have two relevant branches:
fdroid: the release branch for fdroid. almost identical to main, but with firebase removed since it's not opensource.
webrtc: very old and has a stalled effort of embedding a stun/turn server

I appreciate any help with the migration towards typescript. thank you.

**Comment by lil5** [comment]
`rnr -rf '(.*)\.jsx$' '$1.tsx' ./src ./test`

rnr has been a huge help

**Comment by lil5** [comment]
1. You've made it impossible to separate the frontend from the backend an `/api` prefix would have been great. This would allow for a Vite dev server + live reload and a reverse proxy to the backend. 

Thus I'm unable to debug what is going wrong, currently running the frontend requires debugging minimized code. 

2. The absolute paths are a nightmare to work with in typescript/babel/jest/vite at the least use a `~`

---

This where my participation ends, I hope I've given you some helpful advice

**Comment by balzack** [comment]
Yes, I understand. For my reference can you point me to an example of your point 2 and what you would have expected to see instead. Thanks for your advice.

**Comment by lil5** [comment]
Instead of using absolute imports like "components/card.tsx" use "src/components/card.tsx" or "~/src/components/card.tsx" 
simpler still "./../card.tsx"

I attempted to add each dir in src as a base in my configs but it just becomes a buggy mess.

**Comment by balzack** [comment]
Okay I see what you mean. I appreciate all of your advice and will incorporate it as I go forward.

**Comment by balzack** [comment]
Thanks for your other PR, I will test it out when I am back from holidays. 

I have taken your previous comments to heart and am refactoring the entire client source to provide a typescript SDK and have the exiting apps use that SDK. It's about a month away from feature complete.

I know you said you participation had ended, but based on your PR, perhaps you want to look and see if the approach is what you would expect: https://github.com/balzack/databag/tree/sdk/app/sdk

**Comment by lil5** [comment]
Thanks, I'm glad you're up for looking at my PR.

I'm a bit worried about how you're refactoring + moving to typescript + adding an sdk, all at the same time. I'll take a look

**Comment by balzack** [comment]
Yes, it's basically a rewrite of the client code base, using the existing app as reference.  I build a module of the SDK, and then the associated portion of the app. So far I am happy with the result, although I am sure there will be some regressions. 

I wouldn't look in too much detail, as there will be lots of changes before it's V1 complete. I felt your previous high level guidance was very valuable.

refactored web app: https://github.com/balzack/databag/tree/sdk/app/client/web
refactored mobile app: https://github.com/balzack/databag/tree/sdk/app/client/mobile
